src/fides/api/alembic/migrations/versions/5fe01e730171_add_external_id_to_privacy_preference.py

```diff
@@ -0,0 +1,219 @@
+"""add external id to privacy preference
+
+Revision ID: 5fe01e730171
+Revises: 4b2eade4353c
+Create Date: 2024-05-31 17:11:50.566937
+
+"""
+
+import sqlalchemy as sa
+import sqlalchemy_utils
+from alembic import op
+from loguru import logger
+
+# revision identifiers, used by Alembic.
+revision = "5fe01e730171"
+down_revision = "4b2eade4353c"
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    op.add_column(
+        "currentprivacypreferencev2",
+        sa.Column(
+            "external_id",
+            sqlalchemy_utils.types.encrypted.encrypted_type.StringEncryptedType(),
+            nullable=True,
+        ),
+    )
+    op.add_column(
+        "currentprivacypreferencev2",
+        sa.Column("hashed_external_id", sa.String(), nullable=True),
+    )
+    op.add_column(
+        "currentprivacypreferencev2",
+        sa.Column("property_id", sa.String(), nullable=True),
+    )
+    op.drop_constraint(
+        "last_saved_for_email", "currentprivacypreferencev2", type_="unique"
+    )
+    op.drop_constraint(
+        "last_saved_for_fides_user_device", "currentprivacypreferencev2", type_="unique"
+    )
+    op.drop_constraint(
+        "last_saved_for_phone_number", "currentprivacypreferencev2", type_="unique"
+    )
+    op.create_unique_constraint(
+        "last_saved_for_email_per_property_id",
+        "currentprivacypreferencev2",
+        ["email", "property_id"],
+    )
+    op.create_unique_constraint(
+        "last_saved_for_external_id_per_property_id",
+        "currentprivacypreferencev2",
+        ["external_id", "property_id"],
+    )
+    op.create_unique_constraint(
+        "last_saved_for_fides_user_device_per_property_id",
+        "currentprivacypreferencev2",
+        ["fides_user_device", "property_id"],
+    )
+    op.create_unique_constraint(
+        "last_saved_for_phone_number_per_property_id",
+        "currentprivacypreferencev2",
+        ["phone_number", "property_id"],
+    )
```

Comment:
Thanks for this nice write up! I didn't know about Squawk, I use this moving forward. I was thinking I had to schedule downtime for this, I even asked Neville about our downtime allotment. I'll try the recommendation here to create the index concurrently and create the constraint using the index

Reply:
@pattisdr I updated the constraint creation commands to use `CREATE USING INDEX` so that we could run them without downtime. I also moved the constraint creation inside the < 1 million row condition. The post-upgrade scripts in Fidesplus have the new `ix_currentprivacypreferencev2_*` composite indices and the new constraint creation commands.